### PR TITLE
[user-authn] Add ability to set multiple domains for DexAuthenticator

### PR DESCRIPTION
**Pull request by trofimovdals**
## Description

Fix [issue](https://github.com/deckhouse/deckhouse/issues/7846)
Now, we can add several ingress resources(applications in DexAuthenticator) for authorization in client applications using the one dex-deployment in namespace
## Why do we need it, and what problem does it solve?

Each DexAuthenticator resource generated a new dex-deployment

## Changelog entries

```changes
section: user-authn
type: feature
summary: Add ability to set multiple domains for DexAuthenticator
impact_level: default
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @trofimovdals. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @trofimovdals. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @trofimovdals. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @trofimovdals. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @trofimovdals. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @trofimovdals. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by trofimovdals** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.27 11410385335 11410385335-1-con-1-27 /sys/deckhouse-oss/install:pr10200 redos@51.250.1.62

**Comment by trofimovdals** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.27 11419944611 11419944611-1-con-1-27 /sys/deckhouse-oss/install:pr10200 redos@89.169.153.134

**Comment by trofimovdals** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.27 11420400253 11420400253-1-con-1-27 /sys/deckhouse-oss/install:pr10200 redos@51.250.93.255

**Comment by trofimovdals** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.27 11420400253 11420400253-1-con-1-27 /sys/deckhouse-oss/install:pr10200 redos@51.250.93.255

**Comment by trofimovdals** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.27 11420400253 11420400253-1-con-1-27 /sys/deckhouse-oss/install:pr10200 redos@51.250.93.255

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @trofimovdals. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @trofimovdals. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by trofimovdals** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.27 11465543760 11465543760-1-con-1-27 /sys/deckhouse-oss/install:pr10200 redos@89.169.159.32

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @trofimovdals. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @trofimovdals. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @trofimovdals. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @trofimovdals. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @trofimovdals. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @trofimovdals. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @trofimovdals. I've started the workflow for label 'e2e/run/yandex-cloud'...

File: modules/150-user-authn/hooks/get_dex_authenticator_crds_test.go [hunk 1]
```diff
@@ -83,13 +85,16 @@ spec:
...
   "spec": {
-    "applicationDomain": "test",
-    "applicationIngressClassName": "nginx",
-    "sendAuthorizationHeader": false,
-    "nodeSelector":
+    "applications": [
       {
-        "testnode": ""
-      },
+        "domainName": "test",
```
Comment (nabokihms) [inline modules/150-user-authn/hooks/get_dex_authenticator_crds_test.go:90]:
```suggestion
        "domain": "test",
```

File: modules/150-user-authn/hooks/get_dex_authenticator_crds_test.go [hunk 2]
```diff
@@ -124,17 +129,18 @@ data:
...
 kind: DexAuthenticator
 metadata:
   name: test
   namespace: d8-dashboard
   annotations:
     dexauthenticator.deckhouse.io/allow-access-to-kubernetes: "true"
 spec:
-  applicationDomain: test
+  applications:
+  - domainName: test
```
Comment (nabokihms) [inline modules/150-user-authn/hooks/get_dex_authenticator_crds_test.go:141]:
```suggestion
  - domain: test
```

File: modules/150-user-authn/hooks/get_dex_authenticator_crds_test.go [hunk 3]
```diff
@@ -148,8 +154,12 @@ spec:
   "name": "test",
   "namespace": "d8-dashboard",
   "spec": {
-    "applicationDomain": "test",
-    "applicationIngressClassName": "nginx",
+    "applications": [
+      {
+        "domainName": "test",
```
Comment (nabokihms) [inline modules/150-user-authn/hooks/get_dex_authenticator_crds_test.go:159]:
```suggestion
        "domain": "test",
```

File: modules/150-user-authn/hooks/get_dex_authenticator_crds_test.go [hunk 4]
```diff
@@ -178,17 +188,18 @@ data:
...
 kind: DexAuthenticator
 metadata:
   name: test
   namespace: d8-monitoring
   annotations:
     dexauthenticator.deckhouse.io/allow-access-to-kubernetes: "true"
 spec:
-  applicationDomain: test
+  applications:
+  - domainName: test
```
Comment (nabokihms) [inline modules/150-user-authn/hooks/get_dex_authenticator_crds_test.go:200]:
```suggestion
  - domain: test
```

File: modules/150-user-authn/hooks/get_dex_authenticator_crds_test.go [hunk 5]
```diff
@@ -214,8 +225,12 @@ data:
   "name": "test",
   "namespace": "d8-monitoring",
   "spec": {
-    "applicationDomain": "test",
-    "applicationIngressClassName": "nginx",
+    "applications": [
+      {
+        "domainName": "test",
```
Comment (nabokihms) [inline modules/150-user-authn/hooks/get_dex_authenticator_crds_test.go:230]:
```suggestion
        "domain": "test",
```

File: modules/150-user-authn/template_tests/authenticator_test.go [hunk 6]
```diff
@@ -50,14 +50,15 @@ var _ = Describe("Module :: user-authn :: helm template :: dex authenticator", f
     appDexSecret: dexSecret
     cookieSecret: cookieSecret
   spec:
-    applicationDomain: authenticator.example.com
-    applicationIngressCertificateSecretName: test
-    applicationIngressClassName: test
+    applications:
+    - domainName: authenticator.example.com
```
Comment (nabokihms) [inline modules/150-user-authn/template_tests/authenticator_test.go:54]:
```suggestion
    - domain: authenticator.example.com
```

File: modules/150-user-authn/template_tests/authenticator_test.go [hunk 7]
```diff
@@ -75,9 +76,10 @@ var _ = Describe("Module :: user-authn :: helm template :: dex authenticator", f
     cookieSecret: cookieSecret
   allowAccessToKubernetes: true
   spec:
-    applicationDomain: authenticator.com
-    applicationIngressCertificateSecretName: test
-    applicationIngressClassName: test
+    applications:
+    - domainName: authenticator.com
```
Comment (nabokihms) [inline modules/150-user-authn/template_tests/authenticator_test.go:80]:
```suggestion
    - domain: authenticator.com
```

File: modules/150-user-authn/template_tests/authenticator_test.go [hunk 8]
```diff
@@ -75,9 +76,10 @@ var _ = Describe("Module :: user-authn :: helm template :: dex authenticator", f
     cookieSecret: cookieSecret
   allowAccessToKubernetes: true
   spec:
-    applicationDomain: authenticator.com
-    applicationIngressCertificateSecretName: test
-    applicationIngressClassName: test
+    applications:
```
Comment (nabokihms) [inline modules/150-user-authn/template_tests/authenticator_test.go:79]: There is the reason to add more than one application for a single test case to check the render of the array of applications.

File: modules/150-user-authn/templates/dex-authenticator/deployment.yaml [hunk 9]
```diff
@@ -220,4 +222,4 @@ spec:
   {{- if not ($context.Values.global.enabledModules | has "vertical-pod-autoscaler") }}
             {{- include "redis_resources" . | nindent 12 }}
   {{- end }}
-{{- end }}
+{{- end }}
```
Comment (nabokihms) [inline modules/150-user-authn/templates/dex-authenticator/deployment.yaml:225]:
```suggestion
{{- end }}

```

File: modules/150-user-authn/templates/dex-authenticator/ingress.yaml [hunk 10]
```diff
@@ -9,16 +15,16 @@ metadata:
...
-  name: {{ $crd.name }}-dex-authenticator
+  name: {{ $crd.name }}{{ $nameSuffix }}-dex-authenticator
   namespace: {{ $crd.namespace }}
   {{- include "helm_lib_module_labels" (list $context (dict "app" "dex-authenticator")) | nindent 2 }}
 spec:
-  ingressClassName: {{ $crd.spec.applicationIngressClassName }}
+  ingressClassName: {{ $app.ingressClassName }}
   rules:
-  - host: {{ $crd.spec.applicationDomain }}
+  - host: {{ $app.domainName }}
```
Comment (nabokihms) [inline modules/150-user-authn/templates/dex-authenticator/ingress.yaml:27]:
```suggestion
  - host: {{ $app.domain }}
```

File: modules/150-user-authn/templates/dex-authenticator/ingress.yaml [hunk 11]
```diff
@@ -29,45 +35,46 @@ spec:
         path: /dex-authenticator
         pathType: ImplementationSpecific
   {{- if (include "helm_lib_module_https_ingress_tls_enabled" $context ) }}
-    {{- if $crd.spec.applicationIngressCertificateSecretName }}
+    {{- if $app.ingressSecretName }}
   tls:
   - hosts:
-    - {{ $crd.spec.applicationDomain }}
-    secretName: {{ $crd.spec.applicationIngressCertificateSecretName }}
+    - {{ $app.domainName }}
```
Comment (nabokihms) [inline modules/150-user-authn/templates/dex-authenticator/ingress.yaml:41]:
```suggestion
    - {{ $app.domain }}
```

File: modules/150-user-authn/templates/dex-authenticator/ingress.yaml [hunk 12]
```diff
@@ -29,45 +35,46 @@ spec:
...
-  name: {{ $crd.name }}-dex-authenticator-sign-out
+  name: {{ $crd.name }}{{ $nameSuffix }}-dex-authenticator-sign-out
   namespace: {{ $crd.namespace }}
   {{- include "helm_lib_module_labels" (list $context (dict "app" "dex-authenticator")) | nindent 2 }}
 spec:
-  ingressClassName: {{ $crd.spec.applicationIngressClassName }}
+  ingressClassName: {{ $app.ingressClassName }}
   rules:
-  - host: {{ $crd.spec.applicationDomain }}
+  - host: {{ $app.domainName }}
```
Comment (nabokihms) [inline modules/150-user-authn/templates/dex-authenticator/ingress.yaml:60]:
```suggestion
  - host: {{ $app.domain }}
```

File: modules/150-user-authn/templates/dex-authenticator/ingress.yaml [hunk 13]
```diff
@@ -29,45 +35,46 @@ spec:
...
+        path: {{ $app.signOutURL }}
         pathType: ImplementationSpecific
     {{- if (include "helm_lib_module_https_ingress_tls_enabled" $context ) }}
-      {{- if $crd.spec.applicationIngressCertificateSecretName }}
+      {{- if $app.ingressSecretName }}
   tls:
   - hosts:
-    - {{ $crd.spec.applicationDomain }}
-    secretName: {{ $crd.spec.applicationIngressCertificateSecretName }}
+    - {{ $app.domainName }}
```
Comment (nabokihms) [inline modules/150-user-authn/templates/dex-authenticator/ingress.yaml:74]:
```suggestion
    - {{ $app.domain }}
```

File: modules/150-user-authn/templates/dex-authenticator/ingress.yaml [hunk 14]
```diff
@@ -29,45 +35,46 @@ spec:
...
+    secretName: {{ $app.ingressSecretName }}
       {{- end }}
     {{- end }}
   {{- end }}
-{{- end }}
+  {{- end }}
+{{- end }}
```
Comment (nabokihms) [inline modules/150-user-authn/templates/dex-authenticator/ingress.yaml:80]:
```suggestion
{{- end }}

```

File: modules/150-user-authn/templates/dex-authenticator/oauth2client.yaml [hunk 15]
```diff
@@ -17,5 +17,7 @@ allowedGroups:
 {{- end }}
     {{- end }}
 redirectURIs:
-- https://{{ $crd.spec.applicationDomain }}/dex-authenticator/callback
+  {{- range $app := $crd.spec.applications }}
+- https://{{ $app.domainName }}/dex-authenticator/callback
 {{- end }}
+{{- end }}
```
Comment (nabokihms) [inline modules/150-user-authn/templates/dex-authenticator/oauth2client.yaml:23]:
```suggestion
{{- end }}

```

File: modules/150-user-authn/templates/dex-authenticator/secret.yaml [hunk 16]
```diff
@@ -1,6 +1,7 @@
 {{- $context := . }}
 {{ $namespaces := list }}
 {{- range $crd := $context.Values.userAuthn.internal.dexAuthenticatorCRDs }}
+  {{- range $app := $crd.spec.applications }}
```
Comment (nabokihms) [inline modules/150-user-authn/templates/dex-authenticator/secret.yaml:4]:
```suggestion
```

File: modules/150-user-authn/templates/dex-authenticator/secret.yaml [hunk 17]
```diff
@@ -15,7 +16,7 @@ data:
   client-secret: {{ $crd.credentials.appDexSecret | b64enc }}
   {{- end }}
   cookie-secret: {{ $crd.credentials.cookieSecret | b64enc }}
-
+  {{- end }}
```
Comment (nabokihms) [inline modules/150-user-authn/templates/dex-authenticator/secret.yaml:19]:
```suggestion
```

File: modules/150-user-authn/templates/dex-authenticator/secret.yaml [hunk 18]
```diff
@@ -27,6 +28,6 @@ metadata:
 type: kubernetes.io/dockerconfigjson
 data:
   .dockerconfigjson: {{ $context.Values.global.modulesImages.registry.dockercfg }}
-    {{- $namespaces = append $namespaces $crd.namespace }}
+  {{- $namespaces = append $namespaces $crd.namespace }}
   {{- end }}
-{{- end }}
+{{- end }}
```
Comment (nabokihms) [inline modules/150-user-authn/templates/dex-authenticator/secret.yaml:33]:
```suggestion
{{- end }}

```

File: modules/150-user-authn/templates/dex-authenticator/service.yaml [hunk 19]
```diff
@@ -15,4 +15,4 @@ spec:
   selector:
     app: dex-authenticator
     name: {{ $crd.name }}
-{{- end }}
+  {{- end }}
```
Comment (nabokihms) [inline modules/150-user-authn/templates/dex-authenticator/service.yaml:18]:
```suggestion
  {{- end }}

```

File: modules/150-user-authn/webhooks/conversion/dex-authentificator [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+  fi
+}
+
+function __on_conversion::v1alpha1_to_v2alpha1() {
+  if converted=$(context::jq -r '.review.request.objects//[] | map(
+    if .apiVersion == "deckhouse.io/v1alpha1" then
+      .apiVersion = "deckhouse.io/v2alpha1" |
+      .spec.sendAuthorizationHeader = .spec.sendAuthorizationHeader |
+      .spec.applications = [{
+        domainName: .spec.applicationDomain,
```
Comment (nabokihms) [inline modules/150-user-authn/webhooks/conversion/dex-authentificator:118]:
```suggestion
        domain: .spec.applicationDomain,
```

File: modules/150-user-authn/webhooks/conversion/dex-authentificator [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+      toVersion: deckhouse.io/v1
+EOF
+}
+
+function __on_conversion::v2alpha1_to_v1alpha1() {
+  if converted=$(context::jq -r '.review.request.objects//[] | map(
+    if .apiVersion == "deckhouse.io/v2alpha1" then
+      .apiVersion = "deckhouse.io/v1alpha1" |
+      .spec += {
+        applicationDomain: .spec.applications[0].domainName,
```
Comment (nabokihms) [inline modules/150-user-authn/webhooks/conversion/dex-authentificator:51]:
```suggestion
        applicationDomain: .spec.applications[0].domain,
```

File: modules/150-user-authn/webhooks/conversion/dex-authentificator [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+
+function __config__() {
+  cat <<EOF
+configVersion: v1
+kubernetesCustomResourceConversion:
+  - name: v1alpha1_to_v2alpha1
+    crdName: dexauthenticators.deckhouse.io
+    conversions:
+    - fromVersion: deckhouse.io/v1alpha1
+      toVersion: deckhouse.io/v2alpha1
```
Comment (nabokihms) [inline modules/150-user-authn/webhooks/conversion/dex-authentificator:27]:
```suggestion
      toVersion: deckhouse.io/v1
```

File: modules/150-user-authn/webhooks/conversion/dex-authentificator [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+source /shell_lib.sh
+
+function __config__() {
+  cat <<EOF
+configVersion: v1
+kubernetesCustomResourceConversion:
+  - name: v1alpha1_to_v2alpha1
```
Comment (nabokihms) [inline modules/150-user-authn/webhooks/conversion/dex-authentificator:23]:
```suggestion
  - name: v1_to_v2alpha1
```

File: modules/150-user-authn/webhooks/conversion/dex-authentificator [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+    crdName: dexauthenticators.deckhouse.io
+    conversions:
+    - fromVersion: deckhouse.io/v1alpha1
+      toVersion: deckhouse.io/v2alpha1
+  - name: v2alpha1_to_v1alpha1
+    crdName: dexauthenticators.deckhouse.io
+    conversions:
+    - fromVersion: deckhouse.io/v2alpha1
+      toVersion: deckhouse.io/v1alpha1
```
Comment (nabokihms) [inline modules/150-user-authn/webhooks/conversion/dex-authentificator:32]:
```suggestion
  - name: v2alpha1_to_v1
    crdName: dexauthenticators.deckhouse.io
    conversions:
    - fromVersion: deckhouse.io/v2alpha1
      toVersion: deckhouse.io/v1
```

File: modules/150-user-authn/webhooks/conversion/dex-authentificator [hunk 25]
```diff
@@ -0,0 +1,140 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+source /shell_lib.sh
+
+function __config__() {
+  cat <<EOF
+configVersion: v1
+kubernetesCustomResourceConversion:
```
Comment (nabokihms) [inline modules/150-user-authn/webhooks/conversion/dex-authentificator:22]: v1alpha1 <-> v1 <-> v2alpha1

**Review by nabokihms — CHANGES_REQUESTED** [review]
Fix the test case here, add a test with two applications
https://github.com/deckhouse/deckhouse/blob/688ac6231c96c661cba089d5beff68e816d28fd1/modules/150-user-authn/openapi/values.yaml#L74-L107

**Pull request closed without merging**